experimental/virtio/src/console.rs

```diff
@@ -0,0 +1,215 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::queue::{DeviceWriteOnlyQueue, DriverWriteOnlyQueue};
+use alloc::collections::VecDeque;
+use anyhow::Context;
+use rust_hypervisor_firmware_virtio::{
+    device::VirtioBaseDevice,
+    pci::{find_device, VirtioPciTransport},
+};
+
+/// The number of buffer descriptors in each of the queues.
+const QUEUE_SIZE: usize = 16;
+
+/// The size of each of the buffers used by the transmit and receive queues.
+const DATA_BUFFER_SIZE: usize = 4096;
+
+/// The id of the receive queue.
+const RX_QUEUE_ID: u16 = 0;
+
+/// The id of the transmit queue.
+const TX_QUEUE_ID: u16 = 1;
+
+/// The virtio device id for a console device.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-2560001>.
+const DEVICE_ID: u16 = 3;
+
+/// The PCI device id for a virtio console device.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-1020002>.
+const PCI_DEVICE_ID: u16 = 0x1040 + DEVICE_ID;
+
+/// Simple driver implementation a virtio serial/console device that only supports a single port and
+/// no configutration.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-39000010>.
+pub struct Console {
+    /// The base virtio-over-PCI device used for configuration and notification.
+    device: VirtioBaseDevice<VirtioPciTransport>,
+    /// The receive queue.
+    rx_queue: DeviceWriteOnlyQueue<QUEUE_SIZE, DATA_BUFFER_SIZE>,
+    /// The transmit queue.
+    tx_queue: DriverWriteOnlyQueue<QUEUE_SIZE, DATA_BUFFER_SIZE>,
+    /// A temporary buffer to store extra data from a buffer that was not fully read when using
+    /// `read_all`.
+    pending_data: Option<VecDeque<u8>>,
+}
+
+impl Console {
+    /// Finds the virtio console PCI device, initialises the device, and configures the queues.
+    pub fn find_and_configure_device() -> anyhow::Result<Self> {
+        // For now we just scan the first 32 devices on PCI bus 0 to find the first one that matches
+        // the vendor ID and device ID.
+        let pci_device = find_device(super::PCI_VENDOR_ID, PCI_DEVICE_ID)
+            .ok_or_else(|| anyhow::anyhow!("Couldn't find a virtio console device."))?;
+        let transport = VirtioPciTransport::new(pci_device);
+        let device = VirtioBaseDevice::new(transport);
+        let mut result = Self::new(device);
+        result.init()?;
+        // Let the device know there are available buffers in the receiver queue.
+        result.device.notify_queue(RX_QUEUE_ID);
+        Ok(result)
+    }
+
+    /// Reads the next available bytes from the receive queue, if any are available.
```

Comment:
Should we also comment that it notifies the device that buffer is available too after reading?